deps/rabbit/src/rabbit_table.erl

```diff
@@ -389,17 +391,22 @@ definitions() ->
         ++ gm:table_definitions()
         ++ mirrored_supervisor:table_definitions().
 
-maybe_ensure_index_route_table() ->
+%% To allow for rolling upgrades, new Mnesia tables must be created behind feature flags.
+%% However, feature flags are not designed to initialize RabbitMQ.
+%% Therefore, this function is called upon node boot and ensures that:
+%%
+%% 1. For a single node cluster that already ran the feature flag migration function creating
+%% the table, but got subsequently reset, the table must be created again upon node boot.
+%%
+%% 2. For a stand-alone node that already ran the feature flag migration function adding a
+%% table copy and subsequently joins a new cluster, it must add a table copy again upon node boot.
+-spec ensure_feature_flag_tables() -> ok.
+ensure_feature_flag_tables() ->
+    %% Currently, direct_exchange_routing_v2 is the only feature flag that creates a Mnesia table.
     case rabbit_feature_flags:is_enabled(direct_exchange_routing_v2) of
         true ->
-            case rabbit_table:ensure_table_copy(rabbit_index_route, node(), ram_copies) of
-                ok ->
-                    ok;
-                {error, Reason} ->
-                    rabbit_log_feature_flags:warning(
-                      "Failed to add copy of table ~s to node ~p: ~p",
-                      [rabbit_index_route, node(), Reason])
-            end;
+            ok = create(rabbit_index_route, rabbit_binding:index_route_table_definition()),
+            ok = ensure_table_copy(rabbit_index_route, node(), ram_copies);
```

Comment:
To me, the reasoning is incorrect or at least confusing. A feature flag sits on top of a feature, not the opposite.

When implementing a new feature, the feature-specific code is responsible for creating any resources it relies on: tables, files, processes, whatever. Then, and only then, we can talk about a feature flag to allow a RabbitMQ node with that feature-specific code to work with a RabbitMQ node which doesn't. A feature should **never** rely on the feature flag to set anything up for it.

Therefore the notion of "ensure_feature_flag_tables()" is incorrect because feature flags are not responsible for any tables.

In the case of this `direct_exchange_routing_v2` feature, we should have a function taking care of creating whatever table it needs (I don't know which module should own that function because currently there is no function, just code duplicated in at least two places).

That function is called when it is relevant during startup. If an upgrade only consists of creating those tables, then the feature flag migration function simply calls that same function.

This whole feature currently feels like it is designed around its feature flag and it's wrong. When the feature flag becomes required, the migration function will go away and at best the comment becomes wrong, at worst, RabbitMQ breaks.

Reply:
For an alternative implementation of routing that is currently enabled by a feature flag, we cannot realistically enabled it when it is used: that's on the hot message path.

At some point the feature flag will go away as this routing implementation will be the primary one. But we have to see how it behaves in the wild first.

Reply:
Thank you, that's very valuable feedback.
I attempt to provide a better way in https://github.com/rabbitmq/rabbitmq-server/pull/5279 where there is "no special code" anymore: In there, the new table `rabbit_index_route` is treated as any other Mnesia table if feature flag `direct_exchange_routing_v2` is enabled.